vitrina/messages/admin.py

```diff
@@ -28,7 +32,17 @@ def add_view(self, request, form_url='', extra_context=None):
         return super(EmailTemplateAdmin, self).add_view(request)
 
     def change_view(self, request, object_id, extra_content=None):
+        import re
         self.exclude = ('created', 'deleted', 'deleted_on', 'modified_on')
+        email_template = EmailTemplate.objects.filter(id=object_id).first()
+        if request.method == "POST":
+            list_keys = request.POST['template'][
+                        request.POST['template'].find("{") + 1:request.POST['template'].rfind("}")].split()
+            template_keys_from_form = [word for word in list_keys if word.startswith("{") or word.endswith("}")]
+            template_keys_from_form = [re.sub('[^a-zA-Z0-9 \n\.]', '', key.strip(".")) for key in template_keys_from_form]
+            for key in template_keys_from_form:
+                if key not in email_template.email_keys:
+                    raise NotValidKeyException("Not valid key template. Check email template.")
```

Comment:
You can't do this, because Django templates are far more complicated and these regexes just covers small subset of it.

Best way to test templates, is to have a sample context for each template and try to render template, with that sample context. If it renders correctly and converts to markdown correctly, then changes can be saved, otherwise, show error and ask to correct.

And another thing, form validation is done in forms, not in views. Errors should be shown for each form field, not as a system exception.

See [Adding custom validation to the admin](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#admin-custom-validation).